docs: add `@pankod/refine-core` `useSelect` hook (#3462)
refinedev/refine: documentation/docs/api-reference/core/hooks/useSelect/index.md

---
title: useSelect
siderbar_label: useSelect
---

import BasicUsageLivePreview from "./basic-usage-live-preview.md";
import OnSearchLivePreview from "./on-search-live-preview.md";

`useSelect` hook allows you to manage any `select` (like a [Html `<select>` tag](https://www.w3schools.com/tags/tag_select.asp), [React Select](https://react-select.com/home), etc...) component. Since it is designed as headless, It expects you to handle the UI.

This hook uses the `useList` hook for fetching data. [Refer to useList hook for details. →](/docs/api-reference/core/hooks/data/useList/)

:::info

> If you're looking for a complete select library, Refine supports two select libraries out-of-the-box.

Reviewer: ```suggestion
If you're looking for a complete select library, refine supports two select libraries out-of-the-box.
```

---
title: useSelect
siderbar_label: useSelect
source: https://github.com/refinedev/refine/blob/next/packages/core/src/hooks/useSelect/index.ts
---

import BasicUsageLivePreview from "./basic-usage-live-preview.md";
import OnSearchLivePreview from "./on-search-live-preview.md";
import SortLivePreview from "./sort-live-preview.md";
import DefaultValueLivePreview from "./default-value-live-preview.md";

`useSelect` hook allows you to manage any `select` (like a [Html `<select>` tag](https://www.w3schools.com/tags/tag_select.asp), [React Select](https://react-select.com/home), etc...) component. Since it is designed as headless, It expects you to handle the UI.

This hook uses the `useList` hook for fetching data. [Refer to useList hook for details. →](/docs/api-reference/core/hooks/data/useList/)

:::info-tip DERIVATIVES

If you're looking for a complete select library, refine has out-of-the-box support for the libraries below:

> -   refine's `useSelect` (for Headless users) - [Example](/docs/examples/core/useSelect)

Reviewer 2: ```suggestion
```
I think, we can remove it. Because we already tell about useSelect for headless users.

Author: I think remove it, only others @aliemir

-   [Ant Design Select](https://ant.design/components/select) (for Ant Design users) - [Documentation](/docs/api-reference/antd/hooks/field/useSelect) - [Example](/docs/examples/field/useSelect)
-   [Material UI Autocomplete](https://mui.com/material-ui/react-autocomplete) (for Material UI users) - [Documentation](/docs/api-reference/mui/hooks/useAutocomplete)
-   [Mantine Select](https://mantine.dev/core/select/) (for Mantine users) - [Documentation](/docs/api-reference/mantine/hooks/useSelect)

:::

## Basic Usage

Here is a basic example of how to use `useSelect` hook.

<BasicUsageLivePreview />

## Realtime Updates

> When `useSelect` hook is mounted, it passes some parameters (`channel`, `resource` etc.) to the `subscribe` method from the `liveProvider`. 

Reviewer 2: I updated this section like this:
```suggestion
When `useSelect` hook is mounted, it will call the `subscribe` method from the `liveProvider` with some parameters such as `channel`, `resource` etc. It is useful when you want to subscribe to the live updates.
```

It is useful when you want to subscribe to the live updates.

[Refer to the `liveProvider` documentation for more information &#8594](/docs/api-reference/core/providers/live-provider)

## Properties
### `resource` <PropTag required />

It will be passed to the `getList` method from the `dataProvider` as parameter via the `useList` hook. The parameter is usually used as an API endpoint path. It all depends on how to handle the `resource` in the `getList` method. See the [creating a data provider](/docs/api-reference/core/providers/data-provider#creating-a-data-provider) section for an example of how resource are handled.

```tsx
useSelect({
    resource: "categories",
});
```

### `optionLabel` and `optionValue`

Allows you to change the `value` and `label` of your options.  
Default values are `optionLabel = "title"` and `optionValue = "id"`

```tsx

> useSelect<ICategory>({
>     resource: "products",
>     // highlight-next-line
>     optionLabel: "name"
> });

Reviewer 2: ```suggestion
useSelect({
    resource: "products",
    optionLabel: "name",
    optionValue: "product_id"
});
```

```

:::tip

Supports nested properties with option [Object path](https://lodash.com/docs/4.17.15#get) syntax.

> ```tsx
> const { options } = useSelect({
>     resource: "categories",
>     // highlight-start
>     optionLabel: "nested.title",
>     optionValue: "nested.id",
>     // highlight-end
> });
> ```

Reviewer 2: ```suggestion
```tsx
const { options } = useSelect({
    resource: "categories",
    optionLabel: "nested.title",
    optionValue: "nested.id"
});
```
```

```tsx
useSelect({
    filter: [
        {
            field: "isActive",
            operator: "eq",
            value: true,
        },
    ],
});
```
### `defaultValue`

Allows to make options selected by default. Adds extra options to `<select>` component. In some cases like there are many entries for the `<select>` and pagination is required, `defaultValue` may not be present in the current visible options and this can break the `<select>` component. To avoid such cases, A seperate `useMany` query is sent to the backend with the `defaultValue` and appended to the options of `<select>`, ensuring the default values exist in the current options array. Since it uses `useMany` to query the necessary data, the `defaultValue` can be a single value or an array of values like the following:

```tsx
useSelect({
    defaultValue: 1, // or [1, 2]
});
```

[Refer to the `useMany` documentation for detailed usage. &#8594](/api-reference/core/hooks/data/useMany.md)

### `debounce`

It allows us to `debounce` the `onSearch` function.  

```tsx
useSelect({
    resource: "categories",

>     // highlight-next-line

Reviewer 2: ```suggestion
```

```tsx
useSelect({
    queryOptions: {
        retry: 3,
    },
});
```

### `pagination`

`pagination` will be passed to the `getList` method from the `dataProvider` as parameter. It is used to send pagination query parameters to the API.

#### `current`

You can pass the `current` page number to the `pagination` property.

```tsx
useSelect({
    pagination: {
        current: 2,
    },
});
```

#### `pageSize`

You can pass the `pageSize` to the `pagination` property.

```tsx
useSelect({
    pagination: {
        pageSize: 20,
    },
});
```

### `hasPagination`

`hasPagination` will be passed to the `getList` method from the `dataProvider` as parameter via the `useList` hook. It is used to determine whether to use server-side pagination or not.

```tsx
useSelect({
    hasPagination: false,
});
```

### `defaultValueQueryOptions`

When the `defaultValue` property is given, the `useMany` data hook is called for the selected records. With this property, you can change the options of this query. If not given, the values given in `queryOptions` will be used.

```tsx
const { options } = useSelect({
    resource: "categories",

>     // highlight-start
>     defaultValueQueryOptions: {
>         onSuccess: (data) => {
>             console.log("triggers when on query return on success");
>         },
>     },
>     // highlight-end

Reviewer 2: ```suggestion
    defaultValueQueryOptions: {
        onSuccess: (data) => {
            console.log("triggers when on query return on success");
        },
    },
```

## FAQ

### How do get all the data without pagination?

You may want to get all the data without pagination. In this case you should use the [`hasPagination`](/docs/api-reference/core/hooks/useSelect/#haspagination) prop.

> Don't forget to implement it in the [data provider](/docs/api-reference/core/providers/data-provider/#getlist).

### How can search for options (Autocomplete)?

[`onSearch`](/docs/api-reference/core/hooks/useSelect/#onsearch) is a function that is used to set the search value. It is useful when you want to search for a specific value. A simple example of this is shown below.

<OnSearchLivePreview />

### How can selected the option by default?

In some cases we only have `id`, it may be necessary to show it selected in the selection box. This hook sends the request via [`useMany`](/docs/api-reference/core/hooks/data/useMany/), gets the data and mark as seleted.

<DefaultValueLivePreview />

### How do change the `label` and `value` on options?

[`optionLabel` and `optionValue`](/docs/api-reference/core/hooks/useSelect/#optionlabel-and-optionvalue) are used to change the value of your options.

> The default values are optionsLabel="title" and optionsValue="id".

Reviewer 3: ```suggestion
The default values are `optionsLabel="title"` and `optionsValue="id"`.
```
